### PR TITLE
gnrc_rpl: proper interface handling

### LINKED ISSUE #3970: RPL: Advertise DODAG only over the assigned interface
`gnrc_rpl` seems to multicast DIOs over all interfaces, though `gnrc_rpl_init` expects an interface as parameter and sets the RPL-nodes multicast address only for that interface. An easy fix would be to set the interface in the `gnrc_rpl_dodag_t` struct and use it in `gnrc_rpl_send()` to generate a `gnrc_netif_hdr` that addresses that interface.

### PR DESCRIPTION
**Pull request by cgundogan**
At DODAG creation, the appropriate interface id will be assigned to the DODAG.
This way, it is not necessary to look up the correct interface id at several places. Furthermore, looking up the interface id like it was done before had some flaws when multiple interfaces were present.

fixes #3970

### CODE REVIEW

File: sys/net/gnrc/routing/rpl/gnrc_rpl.c [hunk 1]
```diff
@@ -117,37 +117,45 @@ gnrc_rpl_instance_t *gnrc_rpl_root_init(uint8_t instance_id, ipv6_addr_t *dodag_
...
+
+    assert(icmpv6 != NULL);
 
     ipv6 = gnrc_pktsnip_search_type(icmpv6, GNRC_NETTYPE_IPV6);
+    netif = gnrc_pktsnip_search_type(icmpv6, GNRC_NETTYPE_NETIF);
 
     assert(ipv6 != NULL);
 
+    if (netif) {
+        iface = ((gnrc_netif_hdr_t *) netif->data)->if_pid;
```
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl.c]: please remove whitespace

Comment (cgundogan) [inline sys/net/gnrc/routing/rpl/gnrc_rpl.c]: which whitespace?

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 2]
```diff
@@ -83,11 +86,16 @@ void gnrc_rpl_send(gnrc_pktsnip_t *pkt, ipv6_addr_t *src, ipv6_addr_t *dst, ipv6
         return;
     }
 
-    if (!gnrc_netapi_dispatch_send(GNRC_NETTYPE_IPV6, GNRC_NETREG_DEMUX_CTX_ALL,hdr)) {
+    pkt = hdr;
+
+    hdr = gnrc_netif_hdr_build(NULL, 0, NULL, 0);
+    ((gnrc_netif_hdr_t *) hdr->data)->if_pid = iface;
```
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: please also remove whitespace here

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 3]
```diff
@@ -243,11 +251,13 @@ static bool _gnrc_rpl_check_DIS_validity(gnrc_rpl_dis_t *dis, uint16_t len)
     return false;
 }
 
-void gnrc_rpl_recv_DIS(gnrc_rpl_dis_t *dis, ipv6_addr_t *src, ipv6_addr_t *dst, uint16_t len)
+void gnrc_rpl_recv_DIS(gnrc_rpl_dis_t *dis, kernel_pid_t iface, ipv6_addr_t *src,
+                       ipv6_addr_t *dst, uint16_t len)
 {
     /* TODO handle Solicited Information Option */
     (void) dis;
     (void) len;
```
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: I guess `dis` and `len` do not require to be "touched" anymore

Comment (cgundogan) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: unrelated to this PR (: but you are right. Will be handled in a follow-up to not bloat the diff

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 4]
```diff
@@ -243,11 +251,13 @@ static bool _gnrc_rpl_check_DIS_validity(gnrc_rpl_dis_t *dis, uint16_t len)
...
+    (void) iface;
```
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: please remove the whitespaces between the cast brackets and the casted variable

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 5]
```diff
@@ -937,8 +946,10 @@ static bool _gnrc_rpl_check_DAO_validity(gnrc_rpl_dao_t *dao, uint16_t len)
     return false;
 }
 
-void gnrc_rpl_recv_DAO(gnrc_rpl_dao_t *dao, ipv6_addr_t *src, uint16_t len)
+void gnrc_rpl_recv_DAO(gnrc_rpl_dao_t *dao, kernel_pid_t iface, ipv6_addr_t *src, uint16_t len)
 {
+    (void) iface;
```
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: also here and below

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c (hunk 3 above)
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: ok

**Comment by cgundogan** [comment]
@BytesGalore thanks for the review. Addressed your comments

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c (hunk 4 above)
Comment (BytesGalore) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: ok, I see we have no convention on this (I mixed it up with stuff from work)
Just forget my whitespace terrror

Comment (cgundogan) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: fixed them already (;

**Comment by BytesGalore** [comment]
I see, sorry for the noise :)

**Comment by cgundogan** [comment]
ready to squash?

**Comment by BytesGalore** [comment]
yes please squash.
Don't know if it would be cool, but the `rpl` shell command could show the interface of the DODAG.

**Comment by cgundogan** [comment]
@BytesGalore that's a nice idea. Squashed and added a new commit adding the interface to the rpl output. I chose to display it at `instance` level, because the `dodag` line is getting quite long.

**Comment by BytesGalore** [comment]
nice :+1: ACK
Lets kick the CI

File: sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c [hunk 6]
```diff
@@ -536,7 +536,13 @@ void gnrc_rpl_recv_DIO(gnrc_rpl_dio_t *dio, ipv6_addr_t *src, uint16_t len)
 
         inst->mop = (dio->g_mop_prf >> GNRC_RPL_MOP_SHIFT) & GNRC_RPL_SHIFTED_MOP_MASK;
         inst->of = gnrc_rpl_get_of_for_ocp(GNRC_RPL_DEFAULT_OCP);
-        gnrc_rpl_dodag_init(inst, &dio->dodag_id);
+
+        if (iface == KERNEL_PID_UNDEF) {
+            iface = gnrc_ipv6_netif_find_by_addr(NULL, &ipv6_addr_all_rpl_nodes);
```
Comment (cgundogan) [inline sys/net/gnrc/routing/rpl/gnrc_rpl_control_messages.c]: @BytesGalore I amended a quick change in this line. I apparently did an error while rebasing, so it was using the static initializer for the all_rpl_nodes address. I changed that to use the global var instead as introduced by a former PR

**Comment by cgundogan** [comment]
both CIs agree - GO

**Pull request merged**